### PR TITLE
ARROW-7148: [C++][Dataset] Major API cleanup

### PR DESCRIPTION
**Pull request by fsaintjacques**
- Make use of `Result<T>` instead of Status/output-parameter (fixes ARROW-7147).
- Make use of type forward aliases, e.g. the KlassPtr pattern. This ensure a normalization of the correct pointer type and a shorter types to write.
- Use std::shared_ptr everywhere for normalization.
- Move implicit filter cast into R binding (fixes ARROW-7188).
- Fix type validation issue.
- Force FileSystemDataSource to use an owning pointer for the filesystem (fixes ARROW-7094).
- Make constructors private.
- Delete not implemented format header, i.e. CSV, JSON and Feather.
- Delete unused classes ProjectedRecordBatchReader.
- Fold SimplerScanner into Scanner.
- Rename FileBasedDataFragment to FileDataFragment and
  FileSystemBasedDataSource to FileSystemDataSource.

### CODE REVIEW

File: cpp/src/arrow/dataset/scanner.cc [hunk 1]
```diff
@@ -36,20 +36,21 @@ namespace dataset {
...
 /// \brief GetScanTaskIterator transforms an Iterator<DataFragment> in a
 /// flattened Iterator<ScanTask>.
 static ScanTaskIterator GetScanTaskIterator(DataFragmentIterator fragments,
-                                            std::shared_ptr<ScanContext> context) {
+                                            ScanContextPtr context) {
   // DataFragment -> ScanTaskIterator
   auto fn = [context](std::shared_ptr<DataFragment> fragment,
                       ScanTaskIterator* out) -> Status {
-    return fragment->Scan(context, out);
+    ARROW_ASSIGN_OR_RAISE(*out, fragment->Scan(context));
```
Comment (bkietz) [inline cpp/src/arrow/dataset/scanner.cc]:
```suggestion
    return fragment->Scan(context)).Value(out);
```

File: cpp/src/arrow/dataset/scanner.cc [hunk 2]
```diff
@@ -60,19 +61,18 @@ static ScanTaskIterator GetScanTaskIterator(DataFragmentIterator fragments,
...
 static ScanTaskIterator ProjectAndFilterScanTaskIterator(
-    ScanTaskIterator it, std::shared_ptr<Expression> filter,
+    ScanTaskIterator it, ExpressionPtr filter,
     std::shared_ptr<ExpressionEvaluator> evaluator,
     std::shared_ptr<RecordBatchProjector> projector) {
   // Wrap the scanner ScanTask with a FilterAndProjectScanTask
-  auto wrap_scan_task = [filter, evaluator, projector](
-                            std::unique_ptr<ScanTask> task) -> std::unique_ptr<ScanTask> {
+  auto wrap_scan_task = [filter, evaluator, projector](ScanTaskPtr task) -> ScanTaskPtr {
     return internal::make_unique<FilterAndProjectScanTask>(std::move(task), filter,
```
Comment (bkietz) [inline cpp/src/arrow/dataset/scanner.cc]: it looks a bit funny to have `make_unique` here since you're return type is `shared_ptr`, but it's not at all incorrect

File: r/src/expression.cpp [hunk 3]
```diff
@@ -20,82 +20,72 @@
...
+  return ds::equal(lhs, rhs);
 }
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::ComparisonExpression> dataset___expr__not_equal(
-    const std::shared_ptr<arrow::dataset::Expression>& lhs,
-    const std::shared_ptr<arrow::dataset::Expression>& rhs) {
+ds::ExpressionPtr dataset___expr__not_equal(const ds::ExpressionPtr& lhs,
+                                            const ds::ExpressionPtr& rhs) {
   return arrow::dataset::not_equal(lhs, rhs);
```
Comment (nealrichardson) [inline r/src/expression.cpp]:
```suggestion
  return ds::not_equal(lhs, rhs);
```
and same for the rest, right?

File: r/src/expression.cpp [hunk 4]
```diff
@@ -20,82 +20,72 @@
 #if defined(ARROW_R_WITH_ARROW)
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::FieldExpression> dataset___expr__field_ref(
-    std::string name) {
-  return std::make_shared<arrow::dataset::FieldExpression>(std::move(name));
+ds::ExpressionPtr dataset___expr__field_ref(std::string name) {
```
Comment (nealrichardson) [inline r/src/expression.cpp]: Why are these changed to have the parent class as the return value? Just makes for some mismatch because the R classes still use the subclass. Can we use `Expression` everywhere and reduce that boilerplate?

File: r/src/dataset.cpp [hunk 5]
```diff
@@ -20,116 +20,98 @@
...
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::DataSource> dataset___DSDiscovery__Finish(
-    const std::shared_ptr<arrow::dataset::DataSourceDiscovery>& discovery) {
-  std::shared_ptr<arrow::dataset::DataSource> out;
-  STOP_IF_NOT_OK(discovery->Finish(&out));
-  return out;
+ds::DataSourcePtr dataset___DSDiscovery__Finish(
+    const ds::DataSourceDiscoveryPtr& discovery) {
+  return ValueOrStop(discovery->Finish());
```
Comment (nealrichardson) [inline r/src/dataset.cpp]: Observation: it looks odd/inconsistent to have `STOP_IF_NOT_OK` but then `ValueOrStop`

File: cpp/src/arrow/dataset/scanner_internal.h [hunk 6]
```diff
@@ -37,7 +37,8 @@ static inline RecordBatchIterator FilterRecordBatch(
   auto filter_fn = [filter, evaluator](std::shared_ptr<RecordBatch> in,
                                        std::shared_ptr<RecordBatch>* out) {
     ARROW_ASSIGN_OR_RAISE(auto selection_datum, evaluator->Evaluate(*filter, *in));
-    return evaluator->Filter(selection_datum, in, out);
+    ARROW_ASSIGN_OR_RAISE(*out, evaluator->Filter(selection_datum, in));
```
Comment (bkietz) [inline cpp/src/arrow/dataset/scanner_internal.h]:
```suggestion
    return evaluator->Filter(selection_datum, in)).Value(out);
```

File: r/src/arrow_types.h [hunk 7]
```diff
@@ -240,7 +241,17 @@ std::shared_ptr<arrow::Array> Array__from_vector(SEXP x, SEXP type);
 std::shared_ptr<arrow::RecordBatch> RecordBatch__from_arrays(SEXP, SEXP);
 std::shared_ptr<arrow::RecordBatch> RecordBatch__from_dataframe(Rcpp::DataFrame tbl);
 
+namespace ds = ::arrow::dataset;
+namespace fs = ::arrow::fs;
```
Comment (nealrichardson) [inline r/src/arrow_types.h]: Can you substitute `fs` in `r/src/filesystem.cpp` everywhere, like you did for `dataset.cpp`?

File: cpp/src/arrow/dataset/scanner_test.cc [hunk 8]
```diff
@@ -119,43 +119,44 @@ TEST_F(TestSimpleScanner, ToTable) {
...
+    DataSourceVector sources;
 
     schema_ = schema({
         field("b", boolean()),
         field("i8", int8()),
         field("i16", int16()),
-        field("i32", int16()),
-        field("i64", int16()),
+        field("i32", int32()),
+        field("i64", int64()),
```
Comment (bkietz) [inline cpp/src/arrow/dataset/scanner_test.cc]: :)

**Review by nealrichardson — COMMENTED** [review]
A few observations and suggestions, but mostly I'm happy if the R test suites are happy, which they appear to be.

File: r/src/arrow_types.h [hunk 9]
```diff
@@ -240,7 +241,17 @@ std::shared_ptr<arrow::Array> Array__from_vector(SEXP x, SEXP type);
...
+
 namespace arrow {
+
+template <typename R>
+auto ValueOrStop(R&& result) -> decltype(std::forward<R>(result).ValueOrDie()) {
```
Comment (bkietz) [inline r/src/arrow_types.h]: +1

**Review by bkietz — APPROVED** [review]
LGTM, just a few nits/follow up items

File: cpp/src/arrow/dataset/discovery.cc [hunk 10]
```diff
@@ -72,7 +70,7 @@ Status FileSystemDataSourceDiscovery::Make(fs::FileSystem* fs,
 
       if (options.exclude_invalid_files) {
         bool supported = true;
-        RETURN_NOT_OK(format->IsSupported(FileSource(path, fs), &supported));
+        RETURN_NOT_OK(format->IsSupported(FileSource(path, fs.get()), &supported));
```
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/discovery.cc]: Because rebase (wasn't in master when I branched), I'll fix it :)

File: cpp/src/arrow/dataset/scanner.cc (hunk 2 above)
Comment (fsaintjacques) [reply cpp/src/arrow/dataset/scanner.cc]: Will, fix.

File: r/src/dataset.cpp [hunk 11]
```diff
@@ -20,116 +20,98 @@
 #if defined(ARROW_R_WITH_ARROW)
 
 // [[arrow::export]]
-std::shared_ptr<arrow::dataset::DataSourceDiscovery> dataset___FSDSDiscovery__Make(
-    const std::shared_ptr<arrow::fs::FileSystem>& fs,
-    const std::shared_ptr<arrow::fs::Selector>& selector) {
-  std::shared_ptr<arrow::dataset::DataSourceDiscovery> discovery;
+ds::DataSourceDiscoveryPtr dataset___FSDSDiscovery__Make(
```
[the remark this answers is not in the document]
Comment (bkietz) [reply r/src/dataset.cpp]: We might put this everywhere eventually. For now it's only applied to classes declared in the datasets module (which is why we have `DataSourcePtr` but not `RecordBatchPtr`). I'll write a follow up JIRA for adding similar aliases throughout Arrow

File: r/src/dataset.cpp (hunk 5 above)
Comment (bkietz) [reply r/src/dataset.cpp]: `StopIfNoValue`?

Comment (nealrichardson) [reply r/src/dataset.cpp]: I meant more the capitalization and underscores

Comment (bkietz) [reply r/src/dataset.cpp]: Ah, that's because `ValueOrStop` is a function while `STOP_IF_NOT_OK` is a macro. The distinction is significant enough that in C++ we intentionally use different naming style to make it obvious which things are macros

Comment (nealrichardson) [reply r/src/dataset.cpp]: I understand, just observing that when it comes time to use them here, it doesn't look right.

**Comment by nealrichardson** [comment]
@ursabot crossbow submit test-ubuntu-18.04-docs

File: r/src/dataset.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (fsaintjacques) [reply r/src/dataset.cpp]: I didn't want to start aliasing the "core" classes, e.g. Schema. I could make it only scoped to `arrow::dataset`, this is what gandiva is doing.

File: r/src/dataset.cpp (hunk 5 above)
Comment (fsaintjacques) [reply r/src/dataset.cpp]: I'll create a trivial macro `VALUE_OR_STOP` for consistency. I also found a serious bug in STOP_IF_NOT_OK that I'll fix.

Comment (nealrichardson) [reply r/src/dataset.cpp]: 😱 great, thank you

**Comment by nealrichardson** [comment]
circle-ci/crossbow confirm that ARROW-7188 is fixed in this patch.

**Review by bkietz — APPROVED** [review]

File: r/src/arrow_types.h [hunk 12]
```diff
@@ -47,18 +47,8 @@ struct data {
     if (!(TEST)) Rcpp::stop(MSG); \
   } while (0)
 
-#define STOP_IF_NOT_OK(s) STOP_IF_NOT(s.ok(), s.ToString())
```
Comment (fsaintjacques) [inline r/src/arrow_types.h]: The error is that `s` is evaluated **twice** if an error is encountered. Can be catastrophic if `s` has side-effects. It is also susceptible to operator precedence since it's not protected by parenthesis.

Comment (nealrichardson) [reply r/src/arrow_types.h]: 👍 now that these macros are just aliases for functions, we could remove them right? (Not trying to increase your scope here, just for future consideration)

**Review by nealrichardson — APPROVED** [review]

File: r/src/arrow_types.h (hunk 12 above)
Comment (fsaintjacques) [reply r/src/arrow_types.h]: Correct, I opted to minimize the diff while preserving code look.

**Comment by nealrichardson** [comment]
Appveyor passed on the fork: https://ci.appveyor.com/project/fsaintjacques/arrow/builds/28946505

I'll merge this.

File: cpp/examples/arrow/dataset-parquet-scan-example.cc [hunk 13]
```diff
@@ -65,7 +65,7 @@ struct Configuration {
   std::shared_ptr<ds::Expression> filter = ("total_amount"_ > 1000.0f).Copy();
 } conf;
 
-std::shared_ptr<ds::Dataset> GetDatasetFromPath(fs::FileSystem* fs,
+std::shared_ptr<ds::Dataset> GetDatasetFromPath(fs::FileSystemPtr fs,
```
Comment (emkornfield) [inline cpp/examples/arrow/dataset-parquet-scan-example.cc]: Hmm, I'm not a huge fan of this pattern was this already existing in the code base or was part of this change to explicitly use it?  It seems like it is something that should have possibly been discussed before hand? (aplogies if I missed it).

Comment (emkornfield) [reply cpp/examples/arrow/dataset-parquet-scan-example.cc]: Aliases are also discouraged by the style guide: https://google.github.io/styleguide/cppguide.html#Aliases

Comment (wesm) [reply cpp/examples/arrow/dataset-parquet-scan-example.cc]: Per discussions on the mailing list I also am not a fan of having aliases in the public API
